app/lib/search/updater.dart

```diff
@@ -4,92 +4,65 @@
 
 import 'dart:async';
 
-import 'package:gcloud/db.dart';
 import 'package:logging/logging.dart';
 
-import '../frontend/models.dart';
-import '../shared/mock_scores.dart';
 import '../shared/search_service.dart';
-import '../shared/utils.dart' show sliceList;
+import '../shared/task_scheduler.dart';
 
-Logger _logger = new Logger('pub.search.updater');
+import 'backend.dart';
+import 'index_ducene.dart';
 
-class PackageIndexUpdater {
-  final PackageIndex index;
-  Timer _timer;
-  bool _updating = false;
+Logger _logger = new Logger('pub.search.updater');
 
-  PackageIndexUpdater(this.index);
+class IndexUpdateTaskSource extends PollingTaskSource {
+  final bool isLocalDevMode;
+  bool _isFirst = true;
+  IndexUpdateTaskSource({this.isLocalDevMode: false})
+      : super(new Duration(hours: 1));
 
-  Future update({int limit}) async {
-    if (_updating) {
-      _logger.warning('Update called before previous round completed.');
-      return;
-    }
-    try {
-      _updating = true;
-      await _doUpdate(limit);
-    } catch (e, st) {
-      _logger.severe('Error while updating search index.', e, st);
-    } finally {
-      _updating = false;
+  @override
+  Stream<Task> pollTasks() {
+    final DateTime updatedAfter =
+        _isFirst ? null : new DateTime.now().subtract(new Duration(days: 1));
+    _isFirst = false;
+    Stream<Task> stream = searchBackend
+        .listPackages(updatedAfter: updatedAfter)
+        .map((String package) => new Task(package, null));
+    if (isLocalDevMode) {
+      // limit memory use and faster warm-up time
+      stream = stream.take(300);
     }
+    return stream;
```

Comment:
Why should we re-index all packages updated in the last day every hour (so on average we're doing 24 re-indexing of the same tarball redundantly) ?

Similar to the analysis, we should remember, e.g. based on timestamp, which ones we already  handled, so we only re-index new packages.

All of this complication is due to the fact that `pollTasks()` is called many times, instead of a single `startStreaming()` :-/